### PR TITLE
Reimplement "spy-on" to work with commands

### PR DESCRIPTION
**Pull request by DarwinAwardWinner**
The new implementation preserves the interactive form of its argument,
and warns if the replacement passed to :and-call-fake changes the
interactive form. This ensures that spied commands can still be called
as commands (i.e. via "command-execute").

I haven't added any tests for this yet.

### CODE REVIEW

File: buttercup.el [hunk 1]
```diff
@@ -560,7 +579,12 @@ KEYWORD can have one of the following values:
                                  :return-value return-value
                                  :current-buffer (current-buffer)))
               return-value)))
-    this-spy-function))
+    ;; Add the interactive form from `fake-function', if any
+    (if (interactive-form fake-function)
+        `(lambda (&rest args)
+           ,(interactive-form fake-function)
+           (apply ',this-spy-function args))
```
Comment (jorgenschaefer) [inline buttercup.el]: I think this can be `(,this-spy-function ,@args)`, no?

File: tests/test-buttercup.el [hunk 2]
```diff
@@ -501,17 +501,40 @@
 
 (describe "The Spy "
   (let (test-function)
-    (before-each
+    (before-all
```
Comment (jorgenschaefer) [inline tests/test-buttercup.el]: Why this change? If one of the tests breaks the binding, all tests would fail now – that's not a particularly useful co-dependence. And the setup is not really expensive.

File: buttercup.el [hunk 3]
```diff
@@ -518,28 +518,47 @@ KEYWORD can have one of the following values:
...
-                                     (lambda (&rest args)
-                                       nil)))))
+      the original function.
+
+If the original function was a command, the generated spy will
+also be a command with the same interactive form, unless
+`:and-call-fake' is used, in which case it is the caller's
+responsibility to ensure ARG is a command."
+  (let* ((orig (symbol-function symbol))
+        (orig-intform (interactive-form orig))
```
Comment (jorgenschaefer) [inline buttercup.el]: I think the indentation is wrong here, but that might be github's display.

File: buttercup.el [hunk 4]
```diff
@@ -518,28 +518,47 @@ KEYWORD can have one of the following values:
...
+        (replacement
+         (cl-case
+             keyword
```
Comment (jorgenschaefer) [inline buttercup.el]: I do not have a strong opinion here, but I think the idiomatic elisp macro would be `pcase` … not sure either of them really add much readability over the cond form.

**Comment by jorgenschaefer** [comment]
Very nice, thank you for your work! This is great!

File: buttercup.el (hunk 3 above)
Comment (DarwinAwardWinner) [inline buttercup.el]: I probably forgot to reindent after switching `let` to `let*`. I'll fix it.

File: tests/test-buttercup.el (hunk 2 above)
Comment (DarwinAwardWinner) [inline tests/test-buttercup.el]: It's explained in the commit message: https://github.com/jorgenschaefer/emacs-buttercup/pull/95/commits/653f7edc2f431a0ebb983a8d3624ad7c24b9adae

Basically, some of the tests aren't actually testing anything if you use `before-each`. For example, the "restores the old value after a spec run" will always pass because the `before-each` will restore the old value even if `spy-on` doesn't.

Comment (jorgenschaefer) [inline tests/test-buttercup.el]: Interesting. I tricked myself with the semantics there :-D

I think it'd be good to add a comment regarding this here, to prevent others from assuming a quick fix there.

File: buttercup.el (hunk 1 above)
Comment (DarwinAwardWinner) [inline buttercup.el]: No, because `this-spy-function` is a variable containing a function (i.e. we need `(symbol-value 'this-spy-function)`. And `args` doesn't need un-quoting.

Comment (jorgenschaefer) [inline buttercup.el]: My bad, misread the code. :-)

**Comment by DarwinAwardWinner** [comment]
I just added a failing test: after running a spied-on command via `command-execute`, the spied-on command doesn't match `:to-have-been-called`.

**Comment by DarwinAwardWinner** [comment]
Actually, #103 could probably be solved completely and cleanly by reimplementing `spy-on` using `advice-add` and `advice-remove` from nadvice.el, which know how to handle commands, autoloads, etc. The problem is that nadvice was only added in Emacs 24.4, so you'd have to drop support for 24.3, which is currently included in the list of versions tested on Travis.

I suppose the old advice system could do the job as well, but it would be more difficult. 

In any case, I've squashed this into two commits, one with the fix to the spy test logic, and one with the new, improved `spy-on` that checks for the right number of arguments and works on commands and autoloads. If you're interested in pursuing the advice-based approach, let me know.

**Comment by jorgenschaefer** [comment]
Using nadvice might be a good idea, but as long as package authors might want to use buttercup to test packages for 24.3, that's sadly not an option.

Also some conflicts in this PR I'm afraid :-(

**Comment by DarwinAwardWinner** [comment]
Ok, I resolved the conflicts. It was a trivial conflict, just two `require` lines added in the same place.
